### PR TITLE
Proxy Setup

### PR DESCRIPTION
**Pull request by rafiberlin**
When the server the Piwik website is running on needs a proxy to resolve his own address, adds the ability to setup the proxy .

### CODE REVIEW

File: PiwikTracker.php [hunk 1]
```diff
@@ -1288,6 +1288,25 @@ public function setRequestTimeout($timeout)
     }
 
     /**
+     * If a proxy is needed to look up the address of the Piwik site, set it with this
+     * @param $proxy
+     * @param int $proxyPort
+     */
+    public function setProxy($proxy, $proxyPort = 80){
```
Comment (tsteur) [inline PiwikTracker.php]: In functions do you mind adding the opening bracket of a function `{` in a new line?

File: PiwikTracker.php [hunk 2]
```diff
@@ -1288,6 +1288,25 @@ public function setRequestTimeout($timeout)
...
+        $this->proxy = $proxy;
+        $this->proxyPort = $proxyPort;
+    }
+
+    public function getProxy(){
+        if(isset($this->proxy) && isset($this->proxyPort)){
```
Comment (tsteur) [inline PiwikTracker.php]: Do you mind adding some spaces between `if(` eg `if (` and before the opening bracket eg `)) {`.

`if (isset($this->proxy) && isset($this->proxyPort)) {`

File: PiwikTracker.php [hunk 3]
```diff
@@ -1288,6 +1288,25 @@ public function setRequestTimeout($timeout)
...
+     * If a proxy is needed to look up the address of the Piwik site, set it with this
+     * @param $proxy
+     * @param int $proxyPort
+     */
+    public function setProxy($proxy, $proxyPort = 80){
+        $this->proxy = $proxy;
+        $this->proxyPort = $proxyPort;
+    }
+
+    public function getProxy(){
```
Comment (tsteur) [inline PiwikTracker.php]: I wonder if we need this method public or if we can have it private? If it's public can you maybe add a little description and phpdoc?

File: PiwikTracker.php [hunk 4]
```diff
@@ -1288,6 +1288,25 @@ public function setRequestTimeout($timeout)
     }
 
     /**
+     * If a proxy is needed to look up the address of the Piwik site, set it with this
+     * @param $proxy
+     * @param int $proxyPort
```
Comment (tsteur) [inline PiwikTracker.php]: Can you maybe add an example and the type for `$proxy`? Just to let everyone easily know in which format something needs to be set. Port is pretty obvious but maybe there could be one or two examples for proxy

Comment (rafiberlin) [inline PiwikTracker.php]: Hi, sorry for the late answer, I have been on parental leave for several months on I haven't checked any work related accounts :)
I have added the descriptions and the code indentations as requested.

File: PiwikTracker.php [hunk 5]
```diff
@@ -1288,6 +1288,30 @@ public function setRequestTimeout($timeout)
...
+    {
+        $this->proxy = $proxy;
+        $this->proxyPort = $proxyPort;
+    }
+
+    /**
+     * If the proxy IP and the proxy port have been set, with the setProxy() function
+     * returns a string, like "173.234.92.107:80"
+     */
+    public function getProxy()
```
Comment (tsteur) [inline PiwikTracker.php]: Hey, thanks for your updated pull request! I'm wondering if this method needs to be actually public? Can we maybe make it `private`?

File: PiwikTracker.php [hunk 6]
```diff
@@ -1288,6 +1288,30 @@ public function setRequestTimeout($timeout)
...
+    {
+        if (isset($this->proxy) && isset($this->proxyPort)) {
+            return $this->proxy.":".$this->proxyPort;
```
Comment (tsteur) [inline PiwikTracker.php]: Can you maybe add some whitespace before and after the dots? That would be great :) Not super important though as PR looks good in general

File: PiwikTracker.php [hunk 7]
```diff
@@ -1326,6 +1351,12 @@ protected function sendRequest($url, $method = 'GET', $data = null, $force = fal
                     'Accept-Language: ' . $this->acceptLanguage
                 ));
 
+
+
+            if($proxy != null){
```
Comment (tsteur) [inline PiwikTracker.php]: Also here would be ideally a whitespace before and after bracket :)

File: PiwikTracker.php [hunk 8]
```diff
@@ -1362,6 +1393,10 @@ protected function sendRequest($url, $method = 'GET', $data = null, $force = fal
                 )
             );
 
+            if($proxy != null){
```
Comment (tsteur) [inline PiwikTracker.php]: same here re whitespace :)

File: PiwikTracker.php [hunk 9]
```diff
@@ -1362,6 +1393,10 @@ protected function sendRequest($url, $method = 'GET', $data = null, $force = fal
...
+                $stream_options['http']['proxy'] = $proxy;
```
Comment (tsteur) [inline PiwikTracker.php]: Have you tested this one? Does it need a protocol like `tcp://$ip..`?

**Comment by tsteur** [comment]
PR looks good in general apart from some minor whitespace etc. Maybe you could check whether it actually works with proxy using PHP streams? We will then merge this for Piwik 2.16.2

**Comment by tsteur** [comment]
It will be good to have this feature in the tracker :+1:

**Comment by rafiberlin** [comment]
Sorry tsteur, I am no longer in the company, where I created the pull request in the first place, I can't test it with php streams or with other protocols like tcp...

And sorry for the formatting, I used to develop with PHPStorm setup like Eclipse...

File: PiwikTracker.php (hunk 5 above)
Comment (rafiberlin) [inline PiwikTracker.php]: Truth be told, I didn't even use the getProxy() function in the project. It's just that I always create it with the same visibility as the set() Method it corresponds to...
Up to you!

Comment (tsteur) [inline PiwikTracker.php]: Sweet, let's make it private then

**Comment by tsteur** [comment]
Do you think you can complete the PR? Otherwise we would close it for now and at some point use this PR as a starting point to implement it. Or maybe someone else would be keen to finish it.

**Comment by rafiberlin** [comment]
Sorry for the late answer! I don't use my github account this days because I don't work with open source project anymore in my new company...

Anyway, I have corrected the last cosmetic issues

**Comment by tsteur** [comment]
Thanks :+1: I'll try to merge

**Comment by tsteur** [comment]
Merged, thank you 👍

**Pull request merged**